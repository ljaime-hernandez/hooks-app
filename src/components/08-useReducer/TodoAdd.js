import React from 'react'
import { useForm } from '../../hooks/useForm'

export const TodoAdd = ({handleAddTodo}) => {

    // the useForm with an empty string helps us to reset the form input box
    const [{description}, handleInputChange, reset] =  useForm({
        description: ''
    })

    const handleSubmit = (e) => {
        e.preventDefault();
        if(description.trim().length <= 1){
            return;
        }

        // adds a Date and time to the id property for the code to create a unique
        // number, the description will be set as the value entered in the input
        // box and the done value is set as false as this is a to-do list which
        // will have a visual toggle effect for whenever the task is done
        const newTodo = {
            id: new Date().getTime(),
            description: description,
            done: false
        };

        // we send the new object to the handleAddTodo function returned in this component
        // as a reference for us to include the adding logic in here instead on the previous
        // component, for optimization purposes
        handleAddTodo(newTodo);
        reset();
    }

    return (
        <>
        <h4>Add task</h4>
        <form onSubmit={handleSubmit}>
            <input
                type="text"
                name="description"
                className="form-control"
                placeholder="Add new task"
                autoComplete="off"
                value={description}
                onChange={handleInputChange}
            />
            <button
                type="submit"
                className='btn btn-outline-primary mt-1 col-12'
            >
                Add
            </button>
        </form>
        </>
    )
}
