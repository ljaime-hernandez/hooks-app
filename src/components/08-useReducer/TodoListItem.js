import React from 'react'

export const TodoListItem = ({todo, index, handleDelete, handleToggle}) => {

    return (
        <li
            key={todo.id}
            className='list-group-item'
        >
            <p 
                 /*used a ternary operation for the element to change its class
                    to empty if the 'done' attribute of the task object is false,
                     but if true then it will set the class as 'complete', customized
                 in the styles.css file */
                className={`${todo.done && 'complete'}`}
                /*sends the id of the object as reference for the toggling of this element
                   to occur if the characteristics are met*/
                onClick={() => handleToggle(todo.id)}
            >
                {index + 1}. {todo.description}</p>
            <button
                className='btn btn-danger'
                value={todo.description}
                /*sends the id of the object as reference for the deleting of this element
                   to occur if the characteristics are met*/
                onClick={() => {handleDelete(todo.id)}}
            >
                Delete
            </button>
        </li>
    )
}
