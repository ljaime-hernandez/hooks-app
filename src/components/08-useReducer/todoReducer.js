export const todoReducer = (state = [], action) => {

    // custom reducer with specific cases arranged in the previous components
    // for the actions to work accordingly. The add type will simply include
    // a new object in the present array of tasks, the delete will return a
    // new array without including the object containing the id calling that
    // function and the toggle will modify one attribute of the task, which is
    // a boolean, from true to false and viceversa 
    switch (action.type) {
        case 'add':
            return [...state, action.payload];
        case 'delete':
            return state.filter(todo => todo.id !== action.payload);
        case 'toggle': 
            return state.map(todo => 
                (todo.id === action.payload)
                ? {...todo, done: !todo.done}
                : todo
                );
            // case 'toggle-old':
            // return state.map(todo => {
            //     if(todo.id === action.payload){
            //         return {
            //             ...todo,
            //             done: !todo.done}
            //         } else {
            //             return todo;
            //         }
            //     }
        default:
            return state;
    }
}