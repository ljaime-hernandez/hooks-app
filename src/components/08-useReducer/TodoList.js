import React from 'react'
import { TodoListItem } from './TodoListItem'

export const TodoList = ({todo, handleDelete, handleToggle}) => {

    return (
        <ul className='list-group list-group-flush'>
                {

                    /*we retrieve the whole list of objects called todo and use a custom component
                        for it to render each piece of information accordingly, we also use the 
                        map method with an index argument which we use to enlist each task, and we
                        also pass the delete and toggle functions as reference */
                    todo.map ((todo, i) => (
                        <TodoListItem 
                            key={todo.id}
                            todo={todo}
                            index={i} 
                            handleDelete={handleDelete} 
                            handleToggle={handleToggle} />
                    ))
                }
        </ul>
    )
}
