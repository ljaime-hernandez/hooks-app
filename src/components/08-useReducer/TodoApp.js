import React, { useEffect, useReducer } from 'react';
import {todoReducer} from './todoReducer';
import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';
import './styles.css';

// as this is an example of a useReducer hook, we can use the init for it to render previous information
// if available. For this component, we are using the localStorage from the browser, which saves the 
// information as strings, but if looked closely, the string is itself a JSON object, so we use the 
// JSON.parse method for it to use it as a json file
const init = () => {
    return JSON.parse(localStorage.getItem('todos')) || [];
}

export const TodoApp = () => {

    // the reducer here uses a custom hook created for us to be able to handle the addition, deletion and
    // toggling of a to-do list, it has an empty array as argument just in case theres nothing saved in the 
    // localStorage and the dispatch function will allow us to understand how the reducer is used
    const [todo, dispatch] = useReducer(todoReducer, [], init);

    // we will use this hook to render the to-do list every time we make changes to the useReducer object,
    // what the useEffect will do is to store the new object in the localStorage, on an item called 'todos'
    // and will transform our json object into a string, as the localStorage stores information as a string
    // and not as an object
    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todo));
    }, [todo])

    // uses the dispatch function sending an object containing the delete string as action and the id for 
    // further identification
    const handleDelete = (todoId) => {
        const action = {
            type: 'delete',
            payload: todoId
        }

        dispatch(action);
    }

    // uses the dispatch function sending an object containing the toggle string as action and the id for 
    // further identification
    const handleToggle = (todoId) => {
        dispatch({
            type: 'toggle',
            payload: todoId
        })
    }

    // we send this function as reference, the logic behind adding a new object is located in the TodoAdd
    // file, this was an optimization done for the code to look cleaner
    const handleAddTodo = (newTodo) => {
        dispatch({
            type: 'add',
            payload: newTodo
        });
    }

    return (
        <div>
            <h1>To-do App </h1>
            <h3>Number of Tasks ({todo.length})</h3>

            {/*send the delete and toggle functions along with the entire list for further actions */}
            <div className='row'>
                <div className='col-7'>
                    <TodoList todo={todo} handleDelete={handleDelete} handleToggle={handleToggle}/>
                </div>
            {/* sends a reference of the add function for further actions*/}
                <div className='col-5'>
                    <TodoAdd handleAddTodo={handleAddTodo}/>
                </div>
            </div>
        </div>
    )
}
