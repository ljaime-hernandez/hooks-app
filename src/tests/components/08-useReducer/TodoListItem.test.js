import { TodoListItem } from "../../../components/08-useReducer/TodoListItem";
import { demoTodos } from "../../fixtures/demoTodos";
import { shallow } from "enzyme";

describe('Tests on TodoItemList component', () => {

    let handleDelete = jest.fn();
    let handleToggle = jest.fn();
    const index = 1;
    let wrapper = shallow(
            <TodoListItem
                todo={demoTodos[1]}
                index={index}
                handleToggle={handleToggle}
                handleDelete={handleDelete} />)

    beforeEach(() => {
        jest.clearAllMocks();
        wrapper = shallow(
            <TodoListItem
                todo={demoTodos[1]}
                index={index}
                handleToggle={handleToggle}
                handleDelete={handleDelete} />);
    });

        test('should render properly', () => {
            
            expect(wrapper).toMatchSnapshot();
        });

         test('should call handleDelete function', () => {

            wrapper.find('button').simulate('click');
            expect(handleDelete).toHaveBeenCalled();
            expect(handleDelete).toHaveBeenCalledWith(demoTodos[1].id);
        });
        
         test('should call handleToggle function', () => {
            
            wrapper.find('p').simulate('click');
            expect(handleToggle).toHaveBeenCalled();
            expect(handleToggle).toHaveBeenCalledWith(demoTodos[1].id);
        });

        test('should show text correctly', () => {
            
            expect(wrapper.find('p').text().trim()).toBe(`${demoTodos[1].id}. ${demoTodos[1].description}`);
        });

        test('should show class complete on p element if the done value is true', () => {
            
            const todo = demoTodos[0];
            todo.done = true;

            wrapper = shallow(
                <TodoListItem
                    todo={todo}
                />);

            expect(wrapper.find('p').hasClass('complete')).toBe(true);
        });


})
