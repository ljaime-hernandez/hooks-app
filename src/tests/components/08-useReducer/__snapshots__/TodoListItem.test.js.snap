// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on TodoItemList component should render properly 1`] = `
<li
  className="list-group-item"
  key="2"
>
  <p
    className="false"
    onClick={[Function]}
  >
    2
    . 
    Learn MongoDB
  </p>
  <button
    className="btn btn-danger"
    onClick={[Function]}
    value="Learn MongoDB"
  >
    Delete
  </button>
</li>
`;
